### PR TITLE
Mask shift value for shift [mem], const

### LINKED ISSUE #19601: RyuJIT: Incorrect 4-byte immediate emitted for shift causes access violation
In debug, RyuJIT seems to try to emit a 4-byte immediate on a `shl` instruction for the following case:
```csharp
public class Program
{
    static ushort s_2;
    static short[] s_5 = new short[]{1};
    static ulong s_8;
    public static void Main()
    {
        var vr2 = s_5[0];
        M9();
    }

    static void M9()
    {
        s_8 <<= (0 & s_2) + 186;
    }
}
```

According to the JIT, this is the disassembly:
```asm
G_M23195_IG03:
       90                   nop
       48C1255C29EEFFBA000000 shl      qword ptr [reloc classVar[0x391252c0]], 186
       90                   nop
```
Of course shifts do not have a 4-byte immediate form, and VS knows better:
```asm
00007FFE51C915DE 90                   nop  
00007FFE51C915DF 48 C1 25 06 36 EE FF BA shl         qword ptr [7FFE51B74BEDh],0BAh  
00007FFE51C915E7 00 00                add         byte ptr [rax],al  
00007FFE51C915E9 00 90 48 8D 65 30    add         byte ptr [rax+30658D48h],dl  
```

This access violates depending on the value of `rax`.

### PR DESCRIPTION
**Pull request by jakobbotsch**
The shift value needs to be masked as otherwise the emitter may end up
believing it needs 4 bytes for constants >= 128. Since an encoding with
4-byte immediate does not exist for shifts, the last part of the
immediate would be interpreted as code and executed.

This was only a problem for constants between 128 and 256 as other
constants are not currently contained.

Fixes #19601

### CODE REVIEW

File: src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -3573,10 +3573,24 @@ void emitter::emitInsRMW(instruction ins, emitAttr attr, GenTreeStoreInd* storeI
...
+        switch (ins)
+        {
+            case INS_rcl_N:
+            case INS_rcr_N:
+            case INS_rol_N:
+            case INS_ror_N:
+            case INS_shl_N:
+            case INS_shr_N:
+            case INS_sar_N:
+                iconVal &= 0x7F;
```
Comment (tannergooding) [inline src/jit/emitxarch.cpp]: I would think we should either support all values (just sign-extend the icon node) or we should mask in a way that matches what the hardware is doing:
> The count is masked to 5 bits (or 6 bits if in 64-bit mode and REX.W = 1).

Just sign-extending the icon node seems like the overall safer option.

Comment (jakobbotsch) [reply src/jit/emitxarch.cpp]: The masking is done like this in many places in this file. If the constants should be sign-extended instead I would prefer to do it consistently in the file.
For example:
https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L3802-L3814

Comment (CarolEidt) [reply src/jit/emitxarch.cpp]: Right, I think it's best to be consistent.

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -13108,7 +13108,7 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                         // The type of the new GT_NEG node cannot just be op1->TypeGet().
                         // Otherwise we may sign-extend incorrectly in cases where the GT_NEG
                         // node ends up feeding directly a cast, for example in
-                        // GT_CAST<ubyte>(GT_MUL(-1, s_1.ubyte)
+                        // GT_CAST<ubyte>(GT_MUL(-1, s_1.ubyte))
```
Comment (CarolEidt) [inline src/jit/morph.cpp]: You've gotta love how the formatting tools will randomly start complaining about something that's been there forever.

Comment (jakobbotsch) [reply src/jit/morph.cpp]: I messed up this comment in #18837 but there was no reason to fix it there and run CI again. I forgot to fix it in the first commit on this PR, but clang was unhappy about the switch missing the default case, so I figured I would fix the comment together with that.

So not the formatting... this time.

**Comment by jakobbotsch** [comment]
I have not checked diffs, though I don't expect there to be any diffs (except for the test case) as any code that hits this will almost surely crash. However I will check and make sure tomorrow.

**Comment by CarolEidt** [comment]
> Also I am curious if the tests/arm/Tests.lst and tests/arm64/Tests.lst still need to be updated? I have checked previous PRs adding tests, and it doesn't seem like they are adding tests to these lists. In this case it is XArch changes anyway, so I suppose it doesn't really matter for ARM or ARM64. But what are the guidelines for updating those files?

I think that we should be adding most tests to those lists, as long as they are not long-running tests, or tests that are known to fail (and for which an issue has been opened). However, since the only actually code change here is to emitxarch.cpp, I'm going to merge this, and perhaps we should open an issue to review the recent test additions and add this and others as needed.

@BruceForstall @jashook - are there guidelines for when to add tests to the arm and arm64 lists?

**Comment by jashook** [comment]
Ideally all tests would be added to the tests.lst file; however, at the moment we are close enough to dropping the smarty dependency that I am ok with the files getting out of sync.

**Pull request merged**